### PR TITLE
Implement subscriptions

### PR DESCRIPTION
**Pull request by dmitrykuzmin**
This PR adds entity/event subscriptions to the Dart client.

The subscriptions are exposed through the `BackendClient.subscribeTo(Topic)` method which returns a holder of `itemAdded`, `itemChanged` and `itemDeleted` streams.

Example is as follows:
```dart
Topic topic = requests.topic().all(TaskListView());
Subscription<TaskListView> subscription = await client.subscribeTo(topic);
subscription.itemChanged.listen(taskListView => updateDisplay(taskListView));
```

The exposed streams propagate all the changes which happen to entities/events and are broadcast, i.e. can be listened to by multiple callers simultaneously.

The subscription is cancelled with the `Subscription.unsubscribe()` method. This causes the server to cancel the subscription and stop reflecting the changes to the Firebase.

--------------

REST Firebase client, for now, does not support the subscriptions.

The integration tests were thus migrated to using the Firebase App based client which works only in web environment. 

As a consequence, the integration tests are now run in the browser instead of the Dart VM.

--------------

The library version advances to `1.2.3`.

### CODE REVIEW

**Comment by dmitrykuzmin** [comment]
@armiol, @dmdashenkov PTAL.

File: client/lib/backend_client.dart [hunk 1]
```diff
@@ -43,12 +45,22 @@ const _contentType = { 'Content-Type': 'application/x-protobuf'};
...
 
-    final String _baseUrl;
+    /// A period which is used to send a "subscription keep-up" request for all active
+    /// subscriptions.
+    ///
+    /// The Spine server cancels the stale subscriptions after some period of time. To prevent this
+    /// from happening to the active subscriptions of this client, we need to periodically send a
+    /// "keep-up" request to the corresponding endpoint.
+    ///
+    static final Duration subscriptionKeepUpPeriod = new Duration(minutes: 2);
```
Comment (dmdashenkov) [inline client/lib/backend_client.dart]: Please make this attribute configurable with a default value.

File: client/lib/firebase_client.dart [hunk 2]
```diff
@@ -32,7 +32,14 @@ class FirebaseClient {
     /// It is expected that the implementations are asynchronous and do not block the calling site
     /// for a long time (i.e. while performing networking).
     ///
-    Stream<String> get(String path) {
-        throw UnimplementedError('FirebaseClient is an interface');
-    }
+    Stream<String> get(String path);
+
+    /// Obtains the "childAdded" event stream of the node under a given path.
+    Stream<String> childAdded(String path);
```
Comment (dmdashenkov) [inline client/lib/firebase_client.dart:44]: Please add doc on the node values. You could reference the doc of `get(..)`.

File: integration-tests/client-test/integration-test/client_test.dart [hunk 3]
```diff
@@ -52,9 +55,48 @@ void main() {
...
+
+            // Listen to the `itemChanged` event.
+            var newTaskName = "";
+            Stream<Task> itemChanged = entitySubscription.itemChanged;
+            itemChanged.listen((task) => newTaskName = task.name);
+
+            // Send the `RenameTask` command.
+            var renameTaskCmd = RenameTask()
+            ..id = taskId
+            ..name = 'New task name';
```
Comment (dmdashenkov) [inline integration-tests/client-test/integration-test/client_test.dart]: Please fix the alignment.

**Review by dmdashenkov — CHANGES_REQUESTED** [review]

**Comment by dmitrykuzmin** [comment]
@dmdashenkov PTAL again.

File: client/lib/backend_client.dart [hunk 4]
```diff
@@ -43,12 +45,27 @@ const _contentType = { 'Content-Type': 'application/x-protobuf'};
...
+    /// The Spine server cancels stale subscriptions after some period of time. To prevent this
+    /// from happening we need to periodically send a "keep-up" request to the corresponding
+    /// endpoint.
+    ///
+    /// This property allows to configure the period with which the request is sent. It should have
+    /// a value at least no less than the subscription life span configured on the server.
+    ///
+    /// The default value is 2 minutes.
+    ///
+    Duration subscriptionKeepUpPeriod = new Duration(minutes: 2);
```
Comment (dmdashenkov) [inline client/lib/backend_client.dart]: Please make it `final` and add a constructor param. Most likely, the  `Duration` ctor is not `const`, thus you'll have to initialize the field manually.

Comment (dmdashenkov) [inline client/lib/backend_client.dart]: The `new` keyword is optional in Dart. Let's not use it.

File: client/lib/spine_client.dart [hunk 5]
```diff
@@ -66,6 +66,10 @@
 ///                       ..who_completed = actorId;
 ///               client.post(requests.command().create(markDone));
 ///           });
+///     // Listen for all `TaskListView` updates and display them.
+///     Subscription<TaskListView> subscription =
+///             await client.subscribeTo(requests.topic().all(TaskListView()));
+///     subscription.itemChanged.listen(taskListView => updateDisplay(taskListView));
```
Comment (dmdashenkov) [inline client/lib/spine_client.dart]: Please place this above the querying. It's a common pattern to subscribe to updates and then query the current state. This way we do not lose any data.

File: client/lib/subscription.dart [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    /// The server will stop reflecting the updates for the topic.
+    ///
+    void unsubscribe() {
+        _closed = true;
+    }
+
+    static Stream<T> checkIsBroadCast<T>(Stream<T> stream) {
+        if (!stream.isBroadcast) {
+            throw new ArgumentError(
+                'All streams passed to an EntitySubscription instance should be broadcast.');
```
Comment (dmdashenkov) [inline client/lib/subscription.dart]:
```suggestion
                'All streams passed to an `EntitySubscription` instance should be broadcast.');
```

File: client/lib/subscription.dart [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+
+    /// Closes this subscription.
+    ///
+    /// The server will stop reflecting the updates for the topic.
+    ///
+    void unsubscribe() {
+        _closed = true;
+    }
+
+    static Stream<T> checkIsBroadCast<T>(Stream<T> stream) {
```
Comment (dmdashenkov) [inline client/lib/subscription.dart]:
```suggestion
    static Stream<T> _checkIsBroadCast<T>(Stream<T> stream) {
```

**Review by dmdashenkov — CHANGES_REQUESTED** [review]

**Comment by dmitrykuzmin** [comment]
@dmdashenkov PTAL again.

**Review by dmdashenkov — APPROVED** [review]

File: client/lib/subscription.dart [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+import 'package:spine_client/spine/web/firebase/subscription/firebase_subscription.pb.dart';
+import 'package:spine_client/src/json.dart';
+import 'package:spine_client/src/known_types.dart';
+
+/// A subscription for event or entity changes.
+///
+/// The [itemAdded], [itemChanged] and [itemRemoved] streams reflect the changes of a corresponding
+/// type.
+///
+/// The streams are broadcast, i.e. can be listened to more than once, and receive all the state
```
Comment (armiol) [inline client/lib/subscription.dart]: Can you please rephrase. I am not sure I understand the English here.

File: integration-tests/client-test/build.gradle [hunk 9]
```diff
@@ -32,7 +32,7 @@ final def integrationTestDir = './integration-test'
 
 task integrationTest(type: Exec) {
     final def pub = 'pub' + (Os.isFamily(Os.FAMILY_WINDOWS) ? '.bat' : '')
-    commandLine pub, 'run', 'test', integrationTestDir
+    commandLine pub, 'run', 'test', integrationTestDir, '-p', 'chrome'
```
Comment (armiol) [inline integration-tests/client-test/build.gradle:38]: Please have a comment on the flag.

**Review by armiol — CHANGES_REQUESTED** [review]
@dmitrykuzmin LGTM in general. Please see my comments.

**Comment by dmitrykuzmin** [comment]
@armiol PTAL again.

**Review by armiol — APPROVED** [review]

**Pull request merged**